### ISSUE TITLE
Feature request: Item update

### ISSUE THREAD

**Issue by Tom-Bom-badil** [smarthomeNG/smarthome#158]
Due to a recent discussion in the forums, I would like to suggest the introduction of a new item attribute "set_item" in the smarthomeNG core.

If set_item is declared, all updates of an item are also brought forward to other target item(s) as specified in set_item. The syntax could be analog to eval_trigger.

Example:
[item1]
  set_item = item2 | item3

Whenever item1 gets a new value, both item2 and item3 will be updated with the same value, too.

As per our forum discussion, in some cases this will help to prevent complex 1-line eval's as well as unnecessary helper items and logics.

Appreciate your review, thank you for all your work!

**Comment by ohinckel** [comment]
Would be helpful if you could also add a link to the discussion to clarify why this is useful (instead of using `eval_trigger = item` ynd `eval = sh.item()`).

**Comment by Tom-Bom-badil** [comment]
In brief: This is mainly for items that can be changed by several other items and evals, and where no central bus system like e.g. KNX can be used as a central register / gateway to listen to the same address on different items (like knx_listen).

My primary example is the level of a ventilation unit ("KWL") - the plugin I am currently developing has by now 6 different items with partly own eval()'s that influence the ventilation level. A 1-line eval() on the level itself, taking in Information from 6 other items, would be too complicated and extremely Long. A separate logic would break the readability of the item structure.

Full Story [here](https://knx-user-forum.de/forum/supportforen/smarthome-py/1015498-eval_update_item-update-auf-fremde-items). We figured out a workaround, but implementation as above improves code readability again.

**Comment by ohinckel** [comment]
So this should be the opposite to `eval_trigger` which configures which item updates will update item too. Where `set_item` will configure which items to update when the item changes. This would ease the configuration and improves the readability.

**Comment by Tom-Bom-badil** [comment]
Exactly. Plus, it would prevent writing unnecessary 'trivial logics' in some special cases.

**Comment by msinn** [comment]
I am going to look into it. But it isn't something that will make it into the release 1.3

**Comment by cstrassburg** [comment]
@msinn I am actually refactoring the Item class and have already implemented a similar feature to this one.

**Comment by Tom-Bom-badil** [comment]
@cstrassburg: Great, thanks a lot again for your work!

A minor note on this: After having the suggestion of smai implemented for a while ([see his eval Statement here](https://knx-user-forum.de/forum/supportforen/smarthome-py/1015498-eval_update_item-update-auf-fremde-items?p=1016065#post1016065)), I found it working pretty well. However, I would still prefer the suggested way as indicated in #1 (or similar, as it probably has been implemented by now already) .

Because the major disadvantage of smai's way is that now the log file is spammed with entries of 'Info' level, telling you that for 'itemx xyz - eval returns None'. Corresponding line in item.py: [here](https://github.com/smarthomeNG/smarthome/blob/master/lib/item.py#L345).

My suggestion would have been to simply put the log level of that line to debug, and the work around would have been fine for most of the use cases (unless you want to update more than just one item, that one is doable but tricky).

Anyways, thanks for considering!

/tom

**Comment by realhik** [comment]
I think this would be a great feature if it had more flexibility to avoid trivial logics. 
E.g. 
```
[itemA]
  on_update = itemB = 1 
  on_change = itemC = False   # only if prev_value != value
  on_change = itemD = sh.itemB()  # multiple use of statement and expression evaluation
  on_change = itemD = sh.itemB()  if value else 0 # expression evaluation
  # multiple use of attribute "on_xxxx" allowed
[itemB]
  ...
[itemC]
  ...
```

**Comment by realhik** [comment]
In fact, it is already possible to alter other items on item updates. Following code is kind of a workaround:
```
      [[[[manuell]]]]
        type = bool
        eval_trigger = .self
        # manual off resets automatic
        eval = value if sh...auto(0) else value
      [[[[auto]]]]
        type = bool
```

**Comment by msinn** [comment]
@realhik You have an Interesting idea for a new syntax:
```
[itemA]
  on_update = itemB = 1 
  on_change = itemC = False   # only if prev_value != value
  on_change = itemD = sh.itemB()  # multiple use of statement and expression evaluation
  on_change = itemD = sh.itemB()  if value else 0 # expression evaluation
  # multiple use of attribute "on_xxxx" allowed
[itemB]
  ...
[itemC]
  ...
```

but multiple identical keys are not interpreted the way you want in the old config format and are not allowed in YAML.

**Comment by realhik** [comment]
Use a delimiter?
```
[itemA]
  on_update = itemB = 1; itemC = False; itemD = sh.itemB()
```

**Comment by msinn** [comment]
Wouldn't make it  very readable, but would work.

**Comment by bmxp** [comment]
I suggest to focus on a yaml compatible syntax as conf is deprecated and support for conf will be dropped for a future release.

**Comment by msinn** [comment]
Since the implementation from @cstrassburg never materialized, I am going to put this issue back onto my agenda.

**Comment by msinn** [comment]
How about the following syntax (derived fromrealhik's suggestion):

```yaml
itemA:
    on_update: itemB = 1           # on every update of itemA
    on_change: itemC = False       # only if prev_value != value
    ...

itemB:
    ...

itemC:
    ...
```

And for update of multiple items use the yaml list-syntax:
```yaml
itemA:
    on_change:
    - itemC = False                       # only if prev_value != value
    - itemD = sh.itemB()                  # multiple use of statement and expression evaluation
    - itemE = sh.itemB() if value else 0  # expression evaluation
    ...

itemB:
    ...

itemC:
    ...
```

**Comment by ohinckel** [comment]
I prefer the suggestion from @msinn (using yaml style and `on_change`, `on_update`). But we should also talk about the value of these options.

Is it something like an expression (like `eval` setting), or is it something special (like `<item-name> = <expression>`)?

If possible we should introduce a common expression language (or just a python expression). This could also provide a standard set of variables (like `item` for the item object, `value` for the value, ...).

Examples:
```yaml
item:
  on_update:
    - itemA = sh.other.item()    # named item assignment (support such expressions?)
    - sh.itemA(sh.otheritem())   # simple python expression (could be standard for all expression, like eval)
    - itemA = ...otherItem       # supporting relative item references?
  eval: sh.itemA() * 100         # eval should use the same expression language (maybe other settings too)
``` 

In the end a standard Python expression would be the best since it
* it standard expression (somebody can lookup how to write them)
* can provide more useful context variables
* is flexible and powerful (one could write everything and we do not need to implement features)
* does not need a special documentation (all different settings are using the same expression language)

**Comment by Tom-Bom-badil** [comment]
Thank's for thinking this further - in the basic stage we didn't think about all these possibilities, but they would be 'pretty cool to have'. Common Python expressions, along with predefined variables, would probably be the easiest way to handle this; although I am not sure if they are also the easiest way to implement the feature.

/tom

**Comment by msinn** [comment]
@ohinckel In your example, the first two list entries are identical in what they do. I this for the *casual* user of snNG the first one is easier to read and understand. I would simple implement a syntax that aloows both:
```
<item> = <expression>   # where <expression> uses exactly the same syntax as eval
<expression>            # is the same, but the expression result is dropped, so you have to 
                        # use sh.itemA(...) to change an item
```
Relative item addressing is supported by the eval-expression, so it would be supported as well.

**Comment by realhik** [comment]
I guess that the most general way and would give great opportunities for item updates!
So the syntax would be:

classic

```
[item1]
  on_update | on_change:
    [<item> =] <expression> | [<item> =] <expression> | ...
```

yaml

```
item1:
    on_update | on_chagne:
    - [<item> =] <expression>
    - [<item> =] <expression>
    - ...
```

**Comment by msinn** [comment]
That is what I described. But the syntax of your CONF example would not work. It would be:
```
[item1]
  on_update = [<item> =] <expression> | [<item> =] <expression> | ...
  on_change = [<item> =] <expression> | [<item> =] <expression> | ...
```

**Comment by realhik** [comment]
Fine! I can only cite Tom-Bom-badil: would be 'pretty cool to have'

**Comment by msinn** [comment]
I can confirm: This feature is going to make it into SmartHomeNG v1.4

**Comment by msinn** [comment]
Item attributes on_update and on_change have been implemented in v1.4.
